File: sdks/csharp/Openakta.Logger/LoggerBuilder.cs
using System.Collections.Generic;
using Openakta.Logger.Sinks;

namespace Openakta.Logger;

public sealed class LoggerBuilder
{
    private string? _service;
    private string? _environment;
    private readonly List<ISink> _sinks = [];
    private Dictionary<string, object?> _defaultContext = [];

    public LoggerBuilder Service(string service)
    {
        _service = service;
        return this;
    }

    public LoggerBuilder Environment(string environment)
    {
        _environment = environment;
        return this;
    }

    public LoggerBuilder AddSink(ISink sink)
    {
        _sinks.Add(sink);
        return this;
    }

    public LoggerBuilder Sinks(IEnumerable<ISink> sinks)
    {
        _sinks.Clear();
        _sinks.AddRange(sinks);
        return this;
    }

    public LoggerBuilder DefaultContext(IDictionary<string, object?> context)
    {
        _defaultContext = new Dictionary<string, object?>(context);
        return this;
    }

    public Logger Build()
    {
        return new Logger(
            ResolveService(_service),
            ResolveEnvironment(_environment),
            _sinks,
            _defaultContext);
    }

    private static string ResolveService(string? candidate)
    {
        var resolved = !string.IsNullOrWhiteSpace(candidate) ? candidate : System.Environment.GetEnvironmentVariable("OPENAKTA_SERVICE");
        if (string.IsNullOrWhiteSpace(resolved))
        {
            throw new InvalidOperationException("Logger requires a service name or OPENAKTA_SERVICE");
        }

        return resolved;
    }

    private static string ResolveEnvironment(string? candidate)
    {
        var resolved = !string.IsNullOrWhiteSpace(candidate) ? candidate : System.Environment.GetEnvironmentVariable("OPENAKTA_ENV");
        return resolved is "production" or "staging" or "development" ? resolved : "production";
    }
}

File: sdks/java/src/main/java/com/openakta/logger/LoggerBuilder.java
package com.openakta.logger;

import com.openakta.logger.sinks.Sink;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class LoggerBuilder {
  private static final ObjectMapper MAPPER = new ObjectMapper();

  private String service;
  private String environment;
  private final List<Sink> sinks = new ArrayList<>();
  private Map<String, Object> defaultContext = new LinkedHashMap<>();

  public LoggerBuilder service(String service) {
    this.service = service;
    return this;
  }

  public LoggerBuilder environment(String environment) {
    this.environment = environment;
    return this;
  }

  public LoggerBuilder addSink(Sink sink) {
    this.sinks.add(sink);
    return this;
  }

  public LoggerBuilder sinks(List<Sink> sinks) {
    this.sinks.clear();
    this.sinks.addAll(sinks);
    return this;
  }

  public LoggerBuilder defaultContext(Map<String, Object> context) {
    this.defaultContext = deepCopy(context);
    return this;
  }

  public Logger build() {
    return new Logger(resolveService(service), resolveEnvironment(environment), sinks, defaultContext);
  }

  private static String resolveService(String candidate) {
    String resolved = candidate != null && !candidate.isBlank() ? candidate : System.getenv("OPENAKTA_SERVICE");
    if (resolved == null || resolved.isBlank()) {
      throw new IllegalStateException("Logger requires a service name or OPENAKTA_SERVICE");
    }
    return resolved;
  }

  private static String resolveEnvironment(String candidate) {
    String resolved = candidate != null && !candidate.isBlank() ? candidate : System.getenv("OPENAKTA_ENV");
    if (resolved == null || resolved.isBlank()) {
      return "production";
    }
    return switch (resolved) {
      case "production", "staging", "development" -> resolved;
      default -> "production";
    };
  }

  @SuppressWarnings("unchecked")
  private Map<String, Object> deepCopy(Map<String, Object> input) {
    return MAPPER.convertValue(input, new TypeReference<LinkedHashMap<String, Object>>() {});
  }
}
